// umi-ui-tasks/lib/ui/components/Dev/index.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireWildcard(require("react"));

var _antd = require("antd");

var _icons = require("@ant-design/icons");

var _uiModule = _interopRequireDefault(require("../../ui.module.less"));

var _enums = require("../../../server/core/enums");

var _util = require("../../util");

var _hooks = require("../../hooks");

var _Analyze = _interopRequireDefault(require("../Analyze"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }

function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(n); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _iterableToArrayLimit(arr, i) { if (typeof Symbol === "undefined" || !(Symbol.iterator in Object(arr))) return; var _arr = []; var _n = true; var _d = false; var _e = undefined; try { for (var _i = arr[Symbol.iterator](), _s; !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }

function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }

var DevComponent = function DevComponent(_ref) {
  var taskType = _ref.taskType,
      namespace = _ref.namespace,
      api = _ref.api,
      _ref$detail = _ref.detail,
      detail = _ref$detail === void 0 ? {} : _ref$detail,
      dispatch = _ref.dispatch,
      dbPath = _ref.dbPath,
      iife = _ref.iife,
      Terminal = _ref.Terminal;
  var intl = api.intl;
  var isEnglish = api.getLocale() === 'en-US';

  var _Form$useForm = _antd.Form.useForm(),
      _Form$useForm2 = _slicedToArray(_Form$useForm, 1),
      form = _Form$useForm2[0];

  var _useState = (0, _react.useState)(false),
      _useState2 = _slicedToArray(_useState, 2),
      modalVisible = _useState2[0],
      setModalVisible = _useState2[1];

  var _useState3 = (0, _react.useState)(''),
      _useState4 = _slicedToArray(_useState3, 2),
      log = _useState4[0],
      setLog = _useState4[1];

  var _useState5 = (0, _react.useState)('log'),
      _useState6 = _slicedToArray(_useState5, 2),
      view = _useState6[0],
      setView = _useState6[1];

  var _useState7 = (0, _react.useState)({
    UMI_UI_SERVER: 'none',
    UMI_UI_PORT: window.location.port,
    BABEL_POLYFILL: true,
    HMR: true,
    BABEL_CACHE: true,
    MOCK: true,
    BROWSER: true,
    CLEAR_CONSOLE: true,
    PORT: null,
    FORK_TS_CHECKER: false,
    UMI_UI: null
  }),
      _useState8 = _slicedToArray(_useState7, 2),
      env = _useState8[0],
      setEnv = _useState8[1];

  var _useInit = (0, _hooks.useInit)(detail),
      _useInit2 = _slicedToArray(_useInit, 1),
      init = _useInit2[0];

  (0, _react.useEffect)(function () {
    if (!init) {
      return function () {};
    }

    if (view === 'log') {
      dispatch({
        type: "".concat(namespace, "/getTaskDetail"),
        payload: {
          taskType: taskType,
          log: true,
          dbPath: dbPath,
          callback: function callback(_ref2) {
            var log = _ref2.log;
            setLog(log);
          }
        }
      });
    }

    if (iife) {
      dev();
    } // UnMount: reset form


    return function () {
      form.resetFields();
      var terminal = (0, _util.getTerminalRefIns)(taskType, api.currentProject.key);

      if (terminal) {
        terminal.clear();
      }
    };
  }, [init, view, iife]);

  function dev() {
    return _dev.apply(this, arguments);
  }

  function _dev() {
    _dev = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee() {
      return regeneratorRuntime.wrap(function _callee$(_context) {
        while (1) {
          switch (_context.prev = _context.next) {
            case 0:
              dispatch({
                type: "".concat(namespace, "/exec"),
                payload: {
                  taskType: taskType,
                  args: {
                    analyze: true,
                    dbPath: dbPath
                  },
                  env: env
                }
              });

            case 1:
            case "end":
              return _context.stop();
          }
        }
      }, _callee);
    }));
    return _dev.apply(this, arguments);
  }

  function cancelDev() {
    return _cancelDev.apply(this, arguments);
  }

  function _cancelDev() {
    _cancelDev = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee2() {
      return regeneratorRuntime.wrap(function _callee2$(_context2) {
        while (1) {
          switch (_context2.prev = _context2.next) {
            case 0:
              dispatch({
                type: "".concat(namespace, "/cancel"),
                payload: {
                  taskType: taskType
                }
              });

            case 1:
            case "end":
              return _context2.stop();
          }
        }
      }, _callee2);
    }));
    return _cancelDev.apply(this, arguments);
  }

  var openModal = function openModal() {
    setModalVisible(true);
  };

  var handleOk = function handleOk() {
    form.validateFields().then(function (values) {
      setEnv(_objectSpread({}, env, {}, values));
      setModalVisible(false);
    }).catch(function (_) {});
  };

  var handleCancel = function handleCancel() {
    setModalVisible(false);
  };

  var toggleView = function toggleView(e) {
    var value = e.target.value;
    setView(value);
  };

  var stopEventPop = function stopEventPop(e) {
    e && e.stopPropagation();
    e && e.preventDefault();
  };

  var EnvLabel = function EnvLabel(props) {
    return /*#__PURE__*/_react.default.createElement("div", null, /*#__PURE__*/_react.default.createElement("div", {
      onClick: stopEventPop,
      className: _uiModule.default.modleLableTitle
    }, intl({
      id: props.title
    })), /*#__PURE__*/_react.default.createElement("div", {
      className: _uiModule.default.modleLableDesc
    }, /*#__PURE__*/_react.default.createElement("span", {
      onClick: stopEventPop
    }, intl({
      id: props.desc
    })), /*#__PURE__*/_react.default.createElement("a", {
      className: _uiModule.default.modleLablelDescIcon,
      href: isEnglish ? props.link.replace(/\/zh\//, '/') : props.link,
      target: "_blank"
    }, intl({
      id: 'org.umi.ui.tasks.env.detail'
    }))));
  };

  var isTaskRunning = detail && [_enums.TaskState.ING, _enums.TaskState.SUCCESS].indexOf(detail.state) > -1;

  var outputRunningInfo = function outputRunningInfo(_ref3) {
    var _map;

    var state = _ref3.state,
        localUrl = _ref3.localUrl,
        hasError = _ref3.hasError;

    if (!state || state === _enums.TaskState.INIT) {
      return null;
    }

    var map = (_map = {}, _defineProperty(_map, _enums.TaskState.ING, {
      status: 'processing',
      text: /*#__PURE__*/_react.default.createElement("span", null, intl({
        id: hasError ? 'org.umi.ui.tasks.dev.state.starting.error' : 'org.umi.ui.tasks.dev.state.starting'
      }))
    }), _defineProperty(_map, _enums.TaskState.SUCCESS, {
      status: 'success',
      text: /*#__PURE__*/_react.default.createElement("span", null, localUrl ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, intl({
        id: 'org.umi.ui.tasks.dev.state.success'
      }), /*#__PURE__*/_react.default.createElement("a", {
        href: localUrl,
        target: "_blank"
      }, localUrl)) : null)
    }), _defineProperty(_map, _enums.TaskState.FAIL, {
      status: 'error',
      text: /*#__PURE__*/_react.default.createElement("span", null, intl({
        id: 'org.umi.ui.tasks.dev.state.fail'
      }))
    }), _map);
    return /*#__PURE__*/_react.default.createElement("div", {
      className: _uiModule.default.runningInfo
    }, /*#__PURE__*/_react.default.createElement(_antd.Badge, {
      status: map[state].status
    }), /*#__PURE__*/_react.default.createElement("span", null, map[state].text));
  };

  var detailHost = "https://umijs.org/".concat(isEnglish ? '' : 'zh');
  return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("h1", {
    className: _uiModule.default.title
  }, intl({
    id: 'org.umi.ui.tasks.dev'
  })), /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_antd.Row, {
    type: "flex",
    justify: "space-between"
  }, /*#__PURE__*/_react.default.createElement(_antd.Col, {
    className: _uiModule.default.buttonGroup
  }, /*#__PURE__*/_react.default.createElement(_antd.Button, {
    size: api.mini ? 'small' : 'default',
    type: "primary",
    onClick: isTaskRunning ? cancelDev : dev
  }, isTaskRunning ? /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_icons.PauseOutlined, null), /*#__PURE__*/_react.default.createElement("span", {
    className: _uiModule.default.runningText
  }, ' ', intl({
    id: 'org.umi.ui.tasks.dev.cancel'
  }))) : /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_icons.CaretRightOutlined, null), /*#__PURE__*/_react.default.createElement("span", {
    className: _uiModule.default.runningText
  }, ' ', intl({
    id: 'org.umi.ui.tasks.dev.start'
  })))), /*#__PURE__*/_react.default.createElement(_antd.Button, {
    size: api.mini ? 'small' : 'default',
    onClick: openModal
  }, intl({
    id: 'org.umi.ui.tasks.envs'
  })), outputRunningInfo(detail), /*#__PURE__*/_react.default.createElement(_antd.Modal, {
    visible: modalVisible,
    title: intl({
      id: 'org.umi.ui.tasks.envs'
    }),
    onOk: handleOk,
    onCancel: handleCancel
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: _uiModule.default.modalContainer
  }, /*#__PURE__*/_react.default.createElement(_antd.Form, {
    name: "devEnv",
    form: form,
    initialValues: env,
    layout: "vertical"
  }, /*#__PURE__*/_react.default.createElement(_antd.Form.Item, {
    label: intl({
      id: 'org.umi.ui.tasks.envs.port'
    }),
    name: "PORT"
  }, /*#__PURE__*/_react.default.createElement(_antd.Input, null)), /*#__PURE__*/_react.default.createElement(_antd.Form.Item, {
    label: /*#__PURE__*/_react.default.createElement(EnvLabel, {
      title: "org.umi.ui.tasks.envs.babelPolyfill",
      desc: "org.umi.ui.tasks.envs.babelPolyfill.desc",
      link: "".concat(detailHost, "/guide/env-variables.html#babel-polyfill")
    }),
    name: "BABEL_POLYFILL",
    valuePropName: "checked"
  }, /*#__PURE__*/_react.default.createElement(_antd.Switch, {
    size: "small"
  })), /*#__PURE__*/_react.default.createElement(_antd.Form.Item, {
    label: /*#__PURE__*/_react.default.createElement(EnvLabel, {
      title: "org.umi.ui.tasks.envs.hmr",
      desc: "org.umi.ui.tasks.envs.hmr.desc",
      link: "".concat(detailHost, "/guide/env-variables.html#hmr")
    }),
    name: "HMR",
    valuePropName: "checked"
  }, /*#__PURE__*/_react.default.createElement(_antd.Switch, {
    size: "small"
  })), /*#__PURE__*/_react.default.createElement(_antd.Form.Item, {
    label: /*#__PURE__*/_react.default.createElement(EnvLabel, {
      title: "org.umi.ui.tasks.envs.babelCache",
      desc: "org.umi.ui.tasks.envs.babelCache.desc",
      link: "".concat(detailHost, "/guide/env-variables.html#babel-cache")
    }),
    name: "BABEL_CACHE",
    valuePropName: "checked"
  }, /*#__PURE__*/_react.default.createElement(_antd.Switch, {
    size: "small"
  })), /*#__PURE__*/_react.default.createElement(_antd.Form.Item, {
    label: /*#__PURE__*/_react.default.createElement(EnvLabel, {
      title: "org.umi.ui.tasks.envs.mock",
      desc: "org.umi.ui.tasks.envs.mock.desc",
      link: "".concat(detailHost, "/guide/env-variables.html#mock")
    }),
    name: "MOCK",
    valuePropName: "checked"
  }, /*#__PURE__*/_react.default.createElement(_antd.Switch, {
    size: "small"
  })), window.g_bigfish ? null : /*#__PURE__*/_react.default.createElement(_antd.Form.Item, {
    label: /*#__PURE__*/_react.default.createElement(EnvLabel, {
      title: "org.umi.ui.tasks.envs.BROWSER",
      desc: "org.umi.ui.tasks.envs.BROWSER.desc",
      link: "".concat(detailHost, "/guide/env-variables.html#browser")
    }),
    name: "BROWSER",
    valuePropName: "checked"
  }, /*#__PURE__*/_react.default.createElement(_antd.Switch, {
    size: "small"
  })), /*#__PURE__*/_react.default.createElement(_antd.Form.Item, {
    label: /*#__PURE__*/_react.default.createElement(EnvLabel, {
      title: "org.umi.ui.tasks.envs.clear",
      desc: "org.umi.ui.tasks.envs.clear.desc",
      link: "".concat(detailHost, "/guide/env-variables.html#clear-console")
    }),
    name: "CLEAR_CONSOLE",
    valuePropName: "checked"
  }, /*#__PURE__*/_react.default.createElement(_antd.Switch, {
    size: "small"
  })), /*#__PURE__*/_react.default.createElement(_antd.Form.Item, {
    label: /*#__PURE__*/_react.default.createElement(EnvLabel, {
      title: "org.umi.ui.tasks.envs.tsCheck",
      desc: "org.umi.ui.tasks.envs.tsCheck.desc",
      link: "".concat(detailHost, "/guide/env-variables.html#fork-ts-checker")
    }),
    name: "FORK_TS_CHECKER",
    valuePropName: "checked"
  }, /*#__PURE__*/_react.default.createElement(_antd.Switch, {
    size: "small"
  })), /*#__PURE__*/_react.default.createElement(_antd.Form.Item, {
    label: /*#__PURE__*/_react.default.createElement(EnvLabel, {
      title: "org.umi.ui.tasks.envs.umiUI",
      desc: "org.umi.ui.tasks.envs.umiUI.desc",
      link: "".concat(detailHost, "/guide/env-variables.html#umi-ui")
    }),
    name: "UMI_UI",
    valuePropName: "checked"
  }, /*#__PURE__*/_react.default.createElement(_antd.Switch, {
    size: "small"
  })))))), /*#__PURE__*/_react.default.createElement(_antd.Col, {
    className: _uiModule.default.formatGroup
  }, /*#__PURE__*/_react.default.createElement(_antd.Radio.Group, {
    size: api.mini ? 'small' : 'default',
    defaultValue: "log",
    value: view,
    buttonStyle: "solid",
    onChange: toggleView
  }, /*#__PURE__*/_react.default.createElement(_antd.Radio.Button, {
    value: "log"
  }, intl({
    id: 'org.umi.ui.tasks.log'
  })), /*#__PURE__*/_react.default.createElement(_antd.Radio.Button, {
    value: "analyze"
  }, intl({
    id: 'org.umi.ui.tasks.analyze'
  }))))), /*#__PURE__*/_react.default.createElement("div", {
    className: _uiModule.default.logContainer
  }, view === 'log' ? /*#__PURE__*/_react.default.createElement(Terminal, {
    onInit: function onInit(ins) {
      if (ins) {
        (0, _util.setTerminalRefIns)(taskType, api.currentProject.key, ins);
      }
    },
    defaultValue: log
  }) : /*#__PURE__*/_react.default.createElement(_Analyze.default, {
    api: api,
    src: detail.analyzePort ? "http://".concat(window.location.hostname, ":").concat(detail.analyzePort) : null
  }))));
};

var _default = DevComponent;
exports.default = _default;
